rest-service/manager_rest/rest/resources_v3_1/nodes.py

```diff
@@ -131,36 +135,71 @@ def post(self):
         with sm.transaction():
             deployment_id = request_dict['deployment_id']
             deployment = sm.get(models.Deployment, deployment_id)
-            nodes = {node.id: node for node in deployment.nodes}
-            self._set_ni_index(sm, deployment, raw_instances)
-            for raw_instance in raw_instances:
-                node_id = raw_instance['node_id']
-                instance = self._instance_from_raw_instance(raw_instance)
-                instance.set_node(nodes[node_id])
-                sm.put(instance)
+            self._prepare_raw_instances(sm, deployment, raw_instances)
+            db.session.execute(
+                models.NodeInstance.__table__.insert(),
+                raw_instances,
+            )
+            db.session.commit()
         return None, 201
 
-    def _instance_from_raw_instance(self, raw_instance):
-        return models.NodeInstance(
-            id=raw_instance['id'],
-            runtime_properties={},
-            state='uninitialized',
-            version=None,
-            relationships=raw_instance.get('relationships') or [],
-            scaling_groups=raw_instance.get('scaling_groups') or [],
-            host_id=raw_instance.get('host_id'),
-            index=raw_instance['index']
-        )
+    def _prepare_raw_instances(self, sm, deployment, raw_instances):
+        if any(item.get('creator') for item in raw_instances):
+            check_user_action_allowed('set_owner')
 
-    def _set_ni_index(self, sm, deployment, raw_instances):
-        existing_instances = sm.list(models.NodeInstance, filters={
-            'deployment_id': deployment.id
-        }, get_all_results=True)
+        existing_instances = sm.list(
+            models.NodeInstance,
+            filters={'deployment_id': deployment.id},
+            include=['index', 'node_id'],
+            get_all_results=True)
         current_node_index = defaultdict(int)
         for ni in existing_instances:
             if ni.index > current_node_index[ni.node_id]:
                 current_node_index[ni.node_id] = ni.index
+
+        nodes = {node.id: node._storage_id for node in deployment.nodes}
+
+        valid_params = ['id', 'runtime_properties', 'state', 'version',
+                        'relationships', 'scaling_groups', 'host_id',
+                        'index', 'visibility',
+                        '_tenant_id', '_node_fk', '_creator_id']
+
+        user_lookup = {}
+
         for raw_instance in raw_instances:
             node_id = raw_instance['node_id']
```

Comment:
you could just `.pop()` here

Reply:
Good point, missed that when I was merging this stuff together.